sycl/source/detail/kernel_bundle_impl.hpp

```diff
@@ -28,20 +28,58 @@ namespace sycl {
 namespace detail {
 
 template <class T> struct LessByHash {
-  bool operator()(const T &LHS, const T &RHS) {
+  bool operator()(const T &LHS, const T &RHS) const {
     return getSyclObjImpl(LHS) < getSyclObjImpl(RHS);
   }
 };
 
+static bool checkAllDevicesAreInContext(const std::vector<device> &Devices,
+                                        const context &Context) {
+  const std::vector<device> &ContextDevices = Context.get_devices();
+  return std::all_of(
+      Devices.begin(), Devices.end(), [&ContextDevices](const device &Dev) {
+        return ContextDevices.end() !=
+               std::find(ContextDevices.begin(), ContextDevices.end(), Dev);
+      });
+}
+
+static bool checkAllDevicesHaveAspect(const std::vector<device> &Devices,
+                                      aspect Aspect) {
+  return std::all_of(Devices.begin(), Devices.end(),
+                     [&Aspect](const device &Dev) { return Dev.has(Aspect); });
+}
+
 // The class is an impl counterpart of the sycl::kernel_bundle.
 // It provides an access and utilities to manage set of sycl::device_images
 // objects.
 class kernel_bundle_impl {
 
+  void common_ctor_checks(bundle_state State) {
+    const bool AllDevicesInTheContext =
+        checkAllDevicesAreInContext(MDevices, MContext);
+    if (MDevices.empty() || !AllDevicesInTheContext)
+      throw sycl::exception(
+          make_error_code(errc::invalid),
+          "Not all devices are associated with the context or "
+          "vector of devices is empty");
```

Comment:
Nit: I think these two cases should be handled individually to make the exception message more helpful.

Reply:
Will consider doing so in a separate PR.